cmd/maya-apiserver/spc-actions/cstorpool_create.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	"github.com/openebs/maya/pkg/storagepool"
+	"fmt"
+)
+// Cas template is a custom resource which has a list of runTasks.
+
+// runTasks are configmaps which has defined yaml templates for resources that needs
+// to be created or deleted for a cstorpool creation or deletion respectively.
+
+// CreateCstorpool is a function that does following:
+// 1. It receives storagepoolclaim object from the spc watcher event handler.
+
+// 2. There are three sources of information for creation of a cstorpool i.e.
+//    storagepoolclaim object,default values in cas template and hardcoded
+//    values in run tasks.
+//    The function will validate the contained information in storagepoolcalim object.
+
+// 3. After successful validation, it will call a worker function for actual cstorpool creation.
+func CreateCstorpool(spcGot *apis.StoragePoolClaim) (error) {
+
+	glog.Infof("Cstorpool create event received for storagepoolclaim %s",spcGot.ObjectMeta.Name)
+
+	// Check wether the spc object has been processed for cstor pool creation
+	if(spcGot.Status.Phase=="Online"){
+		return errors.New("Cstorpool already exists since the status on storagepoolclaim object is Online")
+	}
+
+	// Check for poolType
+	poolType := spcGot.Spec.PoolSpec.PoolType
+	if(poolType==""){
+		return errors.New("Aborting... as no poolType specified")
+	}
+
+	// Check for disks
+	diskList := spcGot.Spec.Disks.DiskList
+	if(len(diskList)==0){
+		return errors.New("Aborting... as no disk specified")
+	}
+
+	// The name of cas template should be provided as annotation in storagepoolclaim yaml
+	// so that it can be used.
+
+	// Check for cas template
+	castTemplateName := spcGot.Annotations[string(v1alpha1.SPCASTemplateCK)]
+	if(castTemplateName==""){
+		return errors.New("Aborting... as no cas template name specified")
+	}
+
+	// Calling worker function to create cstorpool
+	err:=poolCreateWorker(spcGot,castTemplateName)
+
+	if err!=nil{
+		return err
+	}
+
+	return nil
+}
+
+// poolCreateWorker is a worker function which will create a cstorpool
+// successful creation of a cstorpool should involve following successful resource creation:
+// 1. cstorpool ( A custom resource).
+// 2. cstorpool deployment
+// 3. storagepool (A custom resource)
+
+func poolCreateWorker(spcGot *apis.StoragePoolClaim, castTemplateName string) (error) {
+
+	glog.Infof("Creating cstorpool for storagepoolclaim %s via CASTemplate", spcGot.ObjectMeta.Name)
+
+	// Create an empty cstorpool object.
+
+	// This object will be filled in with some of the details present in storagepoolclaim object
+	// that will be required by CAS Engine to create actual cstorpool cr object in kubernetes.
+
+	// As part of building the cstor pool object, some of the details that are getting filled in,
+	// may not be present in actual cstorpool object that lives in kubernetes. These kind of
+	// details are specific to CAS Engine usage only.
+	cstorPool := &v1alpha1.CStorPool{}
```

Comment:
This is not really an object of CR type CStorPool. In fact it is more of a generic type which depicts a pool. maybe, it is better to define a type of CasPool on similar lines as CasVolume and fill it up with all details from SPC object.